common/lockmanager.go

```diff
@@ -0,0 +1,121 @@
+package common
+
+import (
+    "fmt"
+    "sync"
+    "time"
+
+    "github.com/edulinq/autograder/config"
+    "github.com/edulinq/autograder/log"
+)
+
+type lockData struct {
+    timestamp time.Time;
+    mutex sync.RWMutex;
+    isLocked int;
+}
+
+const TICKER_DURATION_HOUR = 1.0;
+
+var (
+    lockManagerMutex sync.Mutex;
+    lockMap sync.Map;
+    ticker *time.Ticker;
+)
+
+func init() {
+    ticker = time.NewTicker(TICKER_DURATION_HOUR * time.Hour);
+    go removeStaleLocks();
+}
+
+func Lock(key string) {
+    lock(key, false);
+}
+
+func Unlock(key string) error {
+    return unlock(key, false);
+}
+
+func ReadLock(key string) {
+    lock(key, true);
+}
+
+func ReadUnlock(key string) error {
+    return unlock(key, true);
+}
+
+func lock(key string, read bool) {
+    lockManagerMutex.Lock();
+
+    val, _ := lockMap.LoadOrStore(key, &lockData{});
+    lock := val.(*lockData);
+
+    // Unlock the lockManagerMutex before acquiring the lock to avoid a deadlock.
+    lockManagerMutex.Unlock();
+
+    if (read) {
+        lock.mutex.RLock();
+    } else {
+        lock.mutex.Lock();
+        lock.isLocked += 1;
```

Comment:
We actually want to track both cases with isLocked.
A lock is still not stale even if only read locks are being acquired.

Also, you can use ++ in go: `lock.isLocked++`.